src/decode.rs

```diff
@@ -0,0 +1,318 @@
+        .transpose()
+        .map_err(|e| format!("failed to read ABI file: {}", e))?;
+    
+    let tvm_msg = ton_sdk::Contract::deserialize_message(&msg_boc[..])
+        .map_err(|e| format!("failed to deserialize message boc: {}", e))?;
+    
+    let mut printer = msg_printer::MsgPrinter::new(&tvm_msg, is_json);
+    let mut result = String::new();
+    let s = &mut result;
+    write!(s, "{}", printer.print(false)).unwrap();
+
+    if abi.is_some() && tvm_msg.body().is_some() {
+        let abi = abi.unwrap();
+        let mut body_vec = Vec::new();
+        serialize_tree_of_cells(&tvm_msg.body().unwrap().into_cell(), &mut body_vec)
+            .map_err(|e| format!("failed to serialize body: {}", e))?;
+        
+        writeln!(s, "{}", print_decoded_body(body_vec, &abi, is_json)?).unwrap();
+    }
+    if is_json { writeln!(s, "}}").unwrap(); }
+    Ok(result)
+}
+
+
+mod msg_printer {
+    use ton_block::*;
+    use ton_types::cells_serialization::serialize_tree_of_cells;
+    use ton_types::Cell;
+    use std::fmt::Write as FmtWrite;
+    
+    pub struct MsgPrinter<'a> {
+        start: &'static str,
+        off: &'static str,
+        end: &'static str,
+        msg: &'a Message,
+        is_json: bool,
+    }
+
+    impl<'a> MsgPrinter<'a> {
+        pub fn new(msg: &'a Message, is_json: bool) -> Self {
+            MsgPrinter {off: " ", start: "\"", end: "\",", msg, is_json }
+        }
+
+        pub fn print(&mut self, close: bool) -> String {
+            let mut result = String::new();
+            let s = &mut result;
+            if self.is_json {
+                write!(s, "{{\n").unwrap();
+            }
+            self.json(s, "Type", &self.print_msg_type());
+            let hdr = self.print_msg_header();
+            self.start = "{\n";
+            self.end = " },";
+            self.off = " ";
+            self.json(s, "Header", &hdr);
+            self.state_init_printer(s);
+            self.start = "\"";
+            if close { self.end = "\""; } else { self.end = "\","; }
+            self.off = " ";
+            self.json(s, "Body", &tree_of_cells_into_base64(
+                self.msg.body().map(|slice| slice.into_cell()).as_ref(),
+            ));
+            if self.is_json && close {
+                write!(s, "}}\n").unwrap();
+            }
+            result
+        }
+        
+        fn print_msg_type(&self) -> String {
+            match self.msg.header() {
+                CommonMsgInfo::IntMsgInfo(_) => "internal",
+                CommonMsgInfo::ExtInMsgInfo(_) => "external inbound",
+                CommonMsgInfo::ExtOutMsgInfo(_) => "external outbound",
+            }.to_owned() + " message"
+        }
+        
+        
+        fn json<T: std::fmt::Display>(&self, s: &mut String, name: &str, value: &T) {
+            write!(s, "{}\"{}\": {}{}{}\n", self.off, name, self.start, value, self.end).unwrap();
+        }
+        
+        fn print_msg_header(&mut self) -> String {
```

Comment:
Maybe better use `ton-block-json` for parsing message to JSON? Then you'll take the same representation as in GraphQL. Some fields as `body`, `data` and `library` may be removed or replaced